diff -aurN a/src/bccsdk.c b/src/bccsdk.c
--- a/src/bccsdk.c	2017-06-16 17:40:31.500380124 +0530
+++ b/src/bccsdk.c	2017-06-16 17:47:27.024373003 +0530
@@ -126,7 +126,7 @@
     return 0;
   for (; s < path; ++s)
   { 
-    if (!(isalnum(*s) || *s == '-' || *s == '.'))
+    if (!(isalnum(*s) || *s == '-' || *s == '.' || *s == '_'))
       return 0;
   }
   return 1;
@@ -145,7 +145,7 @@
 	{
 		/* invalid URL */
 		syslog(LOG_WARNING, "Invalid URL: %s", url);
-		return -1;
+		return -2;
 	}
 	req->serial = serial;
 	req->trusted = trusted != 0;
