
diff --git a/main/src/wl/keymgmt/src/km_key.c b/main/src/wl/keymgmt/src/km_key.c
index a50a45c..a413102 100644
--- a/main/src/wl/keymgmt/src/km_key.c
+++ b/main/src/wl/keymgmt/src/km_key.c
@@ -1,7 +1,7 @@
 /*
  * Implementation of wlc_key operations
  * Copyright (c) 2012-2013 Broadcom Corporation. All rights reserved.
- * $Id: km_key.c 604522 2015-12-07 18:17:15Z $
+ * $Id: km_key.c 688433 2017-03-06 14:09:41Z $
  */
 
 #include "km_key_pvt.h"
@@ -533,6 +533,10 @@ wlc_key_rx_mpdu(wlc_key_t *key, void *pkt, d11rxhdr_t *rxh)
 
 	} else if (rxh->RxStatus1 & RXS_DECERR) {
 		err = BCME_DECERR;
+		if (key->info.algo == CRYPTO_ALGO_TKIP &&
+			body_len <= (DOT11_IV_TKIP_LEN + TKIP_MIC_SIZE + DOT11_ICV_LEN)) {
+			hwdec = FALSE;
+		}
 	}
 
 	if (err == BCME_OK || err == BCME_DECERR) {
